File: core/weapons/homingMissile.js
import { worldSize } from "../utils/constants.js";
import { clamp } from "../utils/math.js";
import { randomNum } from "../utils/random.js";
import { spatial } from "../world/spatialHash.js";
import { toroidalDelta, toroidalDistance, wrap } from "../world/utils.js";
import { world } from "../world/world.js";
import PlasmaCanon from "./plasmaCanon.js";
import Ship from "../player/ships/ship.js";
import Mine from "./mine.js";
import Asteroid from "../world/object/asteroid/asteroid.js";


export default class HomingMissile extends PlasmaCanon {
    constructor({ x, y, angle, ship, color, speed = 10 }) {
        super({
            name: "Homing Missile",
            speed: speed,
            acceleration: 15000,
            x: x,
            y: y,
            width: 20,
            height: 60,
            angle: angle,
            damage: 5,
            range: 25000,
            fireRate: 0.005,
            energyCost: 1000,
            ship,
            color
        });

        this.target = null;
        this.turnRate = 0.01;
    }

    trackEnemy() {
        if (!this.target || this.target?.state === "dead" || this.distanceTraveled <= 200) return;

        // direction to target
        const dx = toroidalDelta(this.target.x, this.x, world.width);
        const dy = toroidalDelta(this.target.y, this.y, world.height);
        // const dx = this.target.x - this.x;
        // const dy = this.target.y - this.y;

        const targetAngle = Math.atan2(dy, dx);

        // smooth rotation toward target
        let diff = (targetAngle + this.angle) + Math.PI / 2;

        // normalize angle (-PI to PI)
        diff = Math.atan2(Math.sin(diff), Math.cos(diff));

        // limit rotation speed
        this.angle += clamp(diff, this.turnRate)
        // this.angle += Math.max(-this.turnRate, Math.min(this.turnRate, diff));
    }

    update(dt, manager) {
        this.trackEnemy(dt);
        super.update(dt, manager);
    }


    closeObject(obj) {
        // if (this.target) return;
        // if (this.ship instanceof EnemyShip && obj instanceof EnemyShip) return;
        // if (this.ship instanceof PlayerShip && obj instanceof PlayerShip) return;
        // else if (!(this.ship instanceof PlayerShip) && !(obj instanceof PlayerShip))
        if (obj instanceof Asteroid) return;
        const targetDistance = this.target ? toroidalDistance(this.target.x, this.target.y, this.x, this.y) : Infinity;
        const newDistance = toroidalDistance(obj.x, obj.y, this.x, this.y);


        this.target = targetDistance > newDistance ? obj : this.target;
    }
}

File: void_front/game/core/weapons/homing_missile.py
from __future__ import annotations

from game.core.utils.math import clamp
from game.core.weapons.plasma_canon import PlasmaCanon
from game.core.world.utils import toroidalDelta, toroidalDistance, wrap
from game.core.world.world import world
import numpy as np

class HomingMissile(PlasmaCanon):
    def __init__(self, options, force=False):
        defaults = {
            "name": "Homing Missile",
            "speed": options.get("speed", 10),
            "acceleration": 15000,
            "x": options.get("x"),
            "y": options.get("y"),
            "width": 15,
            "height": 50,
            "angle": options.get("angle"),
            "damage": 5,
            "range": 25000,
            "fireRate": 0.005,
            "energyCost": 1000,
            "ship": options.get("ship"),
            "color": options.get("color"),
            "img": options.get("img"),
        }
        super().__init__(defaults, force)
        self.target = None
        self.turnRate = 0.01

    def trackEnemy(self):
        if not self.target or getattr(self.target, "state", None) == "dead" or self.distanceTraveled <= 200:
            return
        dx = toroidalDelta(self.target.x, self.x, world.width)
        dy = toroidalDelta(self.target.y, self.y, world.height)
        targetAngle = np.atan2(dy, dx)
        diff = (targetAngle + self.angle) + np.pi / 2
        diff = np.atan2(np.sin(diff), np.cos(diff))
        self.angle += clamp(diff, self.turnRate)

    def update(self, dt, manager=None):
        self.trackEnemy()
        super().update(dt, manager)

    def closeObject(self, obj):
        from game.core.world.object.asteroid.asteroid import Asteroid

        if isinstance(obj, Asteroid):
            return
        targetDistance = toroidalDistance(self.target.x, self.target.y, self.x, self.y) if self.target else float("inf")
        newDistance = toroidalDistance(obj.x, obj.y, self.x, self.y)
        if targetDistance > newDistance:
            self.target = obj
